enhancements/network/multi-network-vrfs.md

```diff
@@ -0,0 +1,499 @@
+
+
+```
+       sysctl -w net.ipv4.tcp_l3mdev_accept=1
+       sysctl -w net.ipv4.udp_l3mdev_accept=1
+```
+
+
+In order to restrict applications from being able to talk on the other networks, it seems like the only choice would be to insert iptables rules to block traffic coming from/to any of the network interfaces for that application. See [https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/Documentation/networking/vrf.rst?h=v6.1](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/Documentation/networking/vrf.rst?h=v6.1) for more details.
+
+From the OVN-Kubernetes perspective, we will support allowing any host networked pod to reach a Kubernetes service. This will work naturally as host network pods will be communicating with the service CIDR. For host network pod to pod direct access across the unique network, OVN-Kubernetes will only support this traffic when the application is bound to the VRF device. Modifying sysctl settings to allow applications to listen across all VRFs will be unsupported. However, Kubelet probes from the host to the a pod primary IP address on the default network will continue to function.
+
+
+### OpenShift Router/Ingress
+
+One of the goals of this effort is to allow ingress services to pods that live on primary unique networks.
+
+OpenShift router is deployed in two different ways:
+
+
+
+* On-prem: deployed as a host networked pod running haproxy, using a keepalived VIP.
+* Cloud: deployed as an OVN networked pod running haproxy, fronted by a cloud external load balancer which forwards traffic to a nodeport service with ETP Local.
+
+OpenShift router is implemented as HAProxy, which is capable of terminating TLS or allowing TLS passthrough and forwarding HTTP connections directly to endpoints of services.
+
+
+#### On-Prem
+
+Due to previously listed constraints in the Host Networked Pods section, host networked pods are unable to talk directly to OVN networked pods unless bound to the VRF itself. HAProxy needs to communicate with potentially many VRFs, so therefore this functionality does not work. However, a host networked pod is capable of reaching any service CIDR. In order to solve this problem, openshift-ingress-operator will be modified so that when it configures HAProxy, it will check if the service belongs to a unique network, and then use the service CIDR as the forwarding path.
+
+
+#### Cloud
+
+In the cloud case, the ovn-networked HAProxy pod cannot reach pods on other networks. Like the On-Prem case, HAProxy will be modified for unique networks to forward to the service CIDR rather than the endpoint directly. The caveat here is that an OVN networked pod only has service access to:
```

Comment:
It is possible to configure a host-network router in the cloud case; it's just not the default.

One of the advantages of a pod-network router is that it integrates with NetworkPolicy better, but it seems like in a multiple-primary-network environment, admins would rather have the router integrate with VRFs better rather than integrating with NetworkPolicy better...

Reply:
Do we allow users to use a host-networked router on any cloud platform?

If we can do an openshift-router instance per network and run that as an ovn-networked pod in the cloud case, then we could directly DNAT to the pod endpoints like it does today and avoid the rest of this workaround logic.